### ISSUE TITLE
Convert hobby plugins to external plugins

### ISSUE THREAD

**Issue by sjwang90** [influxdata/telegraf#9615]
Telegraf's binary size has grown to well over 100MBs. There has been conversation to remove hobby or lesser popular Telegraf plugins to [external plugin](https://github.com/influxdata/telegraf/blob/master/EXTERNAL_PLUGINS.md) to somewhat reduce binary size. We want to make sure there's a precedent of these "hobby" plugins consistently made external plugins. This also reduces the Telegraf maintainer responsibility for these other plugins and allows focus to be on high priority plugins. 

**Recommended list (not final):**
inputs.bcache
inputs.beanstalkd
inputs.csgo
inputs.dcos
inputs.fireboard
inputs.mailchimp
inputs.marklogic
inputs.mcrouter
inputs.minecraft
inputs.neptune_apex
inputs.openldap
inputs.openweathermap
inputs.raindrops
inputs.ravendb
inputs.riak
inputs.teamspeak
inputs.tengine
inputs.trig
inputs.twemproxy
inputs.zipkin
inputs.disque
inputs.leofs
inputs.rethinkdb

If you use one of these plugins and want to advocate that it should stay in core Telegraf, please comment on this issue and share your thoughts.

**Comment by reimda** [comment]
The difficulty with removing these plugins is that we can't just delete them. We will need to open new github projects for each of them. That involves setting up a build against the shim and writing migration docs.

That's more work than leaving them in telegraf.

Most of these lesser used plugins never have any new issues filed against them so we don't spend any time maintaining them in core telegraf.

There is some value in removing them because it reduces the binary size, but I suspect they don't add much to the binary size. Maybe we should make a PR with them disabled, so we can compare the size of the PR built binary with the main release. That will tell us if it's worth removing them for size reasons.

**Comment by reimda** [comment]
See 9618 for the details, but removing these plugins only reduced the amd64 linux binary by 1.65MB. Before removing them it was 125.61MB. That's a 1.3% difference in size.

**Comment by phemmer** [comment]
Being frustrated with the current plugin experience, I opened #9649 to propose a new solution for external plugins. Seems like it would be very relevant to this as it would allow extracting all the built-in plugins without breaking compatibility.

**Comment by srebhan** [comment]
IMO "removing" plugins from this repository is the right approach. I see the following user groups

1. Users caring about the binary size of the project as they need to bring telegraf to small (embedded) devices. Usually those users have very dedicated setups requiring only a small number of input plugins, processors etc.
2. Users that want to _ONLY_ include the plugins they use to reduce the attack surface for security reasons. They naturally want to include their individual plugin-set _ONLY_.
3. Users that don't care about binary size (e.g. due to running this on a powerful machine anyway) and/or want maximum flexibility. Those users want or at least accept builds including all plugins.

Given these groups, you have 1 and 2 requiring _individual_ very specialized plugin sets and 3 potentially wanting all plugins. To reach all mentioned groups you would ultimately need to make _ALL_ plugins "external" plugins. But what does making a plugin "external" mean? Does it mean we exclude it from the build or does it mean we remove it from this code repository or both?

If you look closer from a user perspective, you just want to reduce what is **build** into the binary. As a user you don't care about code structure or repository organization. So the user wants "exclude from build" and probably does not care about "exclude from repository" as long as he or she can grab the plugins easily.

However, from a developer perspective, you should care about the repository and code structure. Removing the plugins from this repository affects the visibility of those plugins. And here I'm not taking about "oh I didn't know it existed" but rather about navigating through the code, realizing effects of "core" code changes, migration/evolution of code, testing, etc. A developer trying to change an API for example would need to pull all "plugin repositories" to migrate them to that new API. This poses a big hurdle to evolve code IMO.

So in my view the ideal constellation would be to move _ALL_ plugins to this repository (i.e. avoid "exclude from repository") and provide a mechanism to achieve "exclude from build". With my PR #8519 you can build individual plugin sets with a minimum of changes to the existing code and maximum backward-compatibility. So user groups 1 and 2 can download the repo, configure their plugin sets and build while InfluxData can provide a "all-batteries-included" binary for user group 3. You might even provide a web-based build-service as you will just need to generate a configuration yaml stating the plugins to build...

**Comment by phemmer** [comment]
> A developer trying to change an API for example would need to pull all "plugin repositories" to migrate them to that new API. This poses a big hurdle to evolve code IMO.

Strongly disagree with this. This is not a new problem, and has been solved since APIs first existed by using versioning.

When an API is created, it should be well thought out, minimal (implements only what is needed without fluff), and extensible so that it can be added onto without breaking what is there already.

In fact what we call "external plugins" (execd) already have an API. The metrics are sent via influx line protocol, and input plugins use signals to perform the dump. That is an API.
That API has been unchanged for years. So fear of the API frequently changing would seem unjustified.

&nbsp;

> I'm not taking about "oh I didn't know it existed" but rather about navigating through the code, realizing effects of "core" code changes, migration/evolution of code, testing, etc.

Yes, this is a significant problem with the current code base. There's so many different "plugins" in the code base that they often step on each other when it comes to shared libraries. Want to update one library because of plugin X? Well that's going to break plugins Y & Z.

&nbsp;

> So in my view the ideal constellation would be to move ALL plugins to this repository 

This repo has almost 1k open issues. About half of them are labeled as feature requests. And a lot of those are requests for new "plugins". I think this makes it pretty clear the current situation is unmaintainable. Making it worse for the sole reason of wanting to avoid a new API doesn't seem like a justified tradeoff.

**Comment by phemmer** [comment]
Oh, there's also a big security aspect to this subject as well. Some plugins need to run as root to function. This means that plugins which don't need root perms have to run as root anyway because they're part of the same telegraf process. Externalizing them so they can run with different permissions is the only way to solve the problem.

**Comment by reimda** [comment]
Thanks for taking a look at this issue @phemmer and @srebhan and for spending time thinking of how to improve telegraf. Since we've hit on the topic of plugin security, I wanted to share an idea we've been working on.

Running each plugin in a separate process has security benefits and is something we want users to have the option of doing. Instead of extracting built-in plugins and running them with execd (or an improved external api like Patrick has in mind) I think we can do it without removing the plugin from telegraf. We've been working on this in a separate branch called `isolateplugins`. It's currently a proof of concept draft PR but if it proves effective and useful we hope to add it to a future telegraf release.
https://github.com/influxdata/telegraf/pull/9668
The PR adds a "single plugin" mode to telegraf that runs a plugin as if it were external, accepting input or producing output over stdio. Single plugin mode is enabled with a command line switch. It also changes telegraf when running normally (not in single plugin mode) to wrap each input and output plugin instance in an execd plugin that then runs the plugin through a separate telegraf process in single plugin mode. The PR currently runs every plugin in a separate process but we'll probably add a config option to let the user choose which plugins to run in separate process, or require only potentially dangerous plugins to run in a separate process.

This may be getting close to what Patrick mentioned ("some solution like launching multiple telegrafs processes via execd & sudo, which is not a good solution") but it is nearly transparent to the user, so hopefully it doesn't have the drawbacks that could make it not a good solution.

If we use this idea to run each plugin in a separate process, we don't need to make a process external to get the security benefits of running the plugin as a different user. We can also use platform-specific security features to tighten down security even more. For example, the isolateplugin branch adds a linux seccomp whitelist to the plugin process.

Take a look at the branch if you're interested. I brought it up here because it could be a good way to add security without forcing us to take each plugin out of telegraf.

**Comment by srebhan** [comment]
Thanks @phemmer for taking the time to discuss this topic! Much appreciated. Anyway, I think you misunderstood some of my points, so let me clarify those...

Disclaimer: I'm not affiliated with InfluxData, just doing reviews as a member of the community team, so this does not necessarily represent the opinion of InfluxData or the telegraf team.

**TL;DR** If you read my comment carefully I'm _not_ against decoupling of the plugins (quite the contrary) but I'm not in favor of splitting the plugins to different git repositories.

> > A developer trying to change an API for example would need to pull all "plugin repositories" to migrate them to that new API. This poses a big hurdle to evolve code IMO.
> 
> Strongly disagree with this. This is not a new problem, and has been solved since APIs first existed by using versioning.
>
> When an API is created, it should be well thought out, minimal (implements only what is needed without fluff), and extensible so that it can be added onto without breaking what is there already.
>
> In fact what we call "external plugins" (execd) already have an API. The metrics are sent via influx line protocol, and input plugins use signals to perform the dump. That is an API.
> That API has been unchanged for years. So fear of the API frequently changing would seem unjustified.


I agree to the point that a stable API is required and as you rightfully stated we do have a stable API for external plugins. However, assuming that you can design an API *perfectly* is utopia and in fact we already see this. While the API you mentioned exists for years it has limitations (e.g. when sending a large number of datapoint over stdin/stdout). It is very likely that we need a new version of that API to overcome those limitations. Now if you stray the plugins all over the place, how do you want to migrate all those plugins to your shiny new API? Of course you can hope for the other repo-owners to adapt but this process is usually slow and you will maintain a zoo of those API versions in telegraf to keep a core set of plugins working.

> Yes, this is a significant problem with the current code base. There's so many different "plugins" in the code base that they often step on each other when it comes to shared libraries. Want to update one library because of plugin X? Well that's going to break plugins Y & Z.

This is strongly exaggerated and I disagree strongly. Yes we did have this in a few cases, but usually the plugins causing those problems do also suffer from a grown structure. Anyway, as I wrote earlier, I'd be happy to decouple the plugins from each other (e.g. making each of them an own process), but I still want to keep them in one central place (aka repository).
 
> > So in my view the ideal constellation would be to move ALL plugins to this repository
> 
> This repo has almost 1k open issues dating back 6 years. About half of them are labeled as feature requests. And a lot of those are requests for new "plugins". I think this makes it pretty clear the current situation is unmaintainable. Making it worse for the sole reason of wanting to avoid a new API doesn't seem like a justified tradeoff.

Well I don't see how this will improve if you split the present repository into a "core" and hundreds of "plugin-xyz" repositories. Right, the number of issues in "core" will go down dramatically and each of the "plugin-xyz" repos also only has a few issues, but in total the number will be similar.

Additionally, where does a user request a new plugin, which most of the "feature-requests" do, in that new structure? How do you guarantee a certain quality level of the code? There has been requests from "external" plugin owners to review their code, so the workload for the Influx-team will not be reduced.
TBH given the size of the telegraf team, those guys do an amazing job in keep the project running. Splitting their effort onto different repositories will not improve things IMO.

All this being said, here is my point. Splitting the repository will not gain us anything, quite the contrary. I don't want to denial a single plugin being part of this repository. I want to have them a proper review and being merged in here if they like to. This explicitly does **not** mean that we should keep telegraf monolithic (I'm open to split it.) or to restrict "external" plugins in any way! Both "internal" and "external" actually should use the same mechanisms/APIs. However, IMO the telegraf then should _not_ make any guarantee on the quality of those external plugins and cannot guarantee their functionality.

In my view the whole code-organization is a question of responsibilities. For the current repository, the telegraf team tries to guarantee the functionality and a certain code quality. If you find issues, you can report them here and they will get fixed hopefully. For "external" plugins, the telegraf team should not give any guarantee whatsoever and the respective maintainer is responsible.
Now of course the current situation, large number of issues and open PRs, is not satisfactory. It might be a good idea to restructure responsibilities, be it "subsystem" maintainers or designated plugin maintainers. However, this requires a certain manpower... There is a community team and there always is the opportunity to contact Influx people if you want to contribute your time. :-)

Looking forward to your answer and your ideas!
